File: 1582-design-browser-history/design-browser-history.cpp
class BrowserHistory {
public:
    vector<string> h;
    int curr;
    BrowserHistory(string homepage) {
       h.push_back(homepage);
       curr=0;
    }
    
    void visit(string url) {
        h.resize(curr+1);
        h.push_back(url);
        curr++;
    }
    
    string back(int steps) {
        curr-=steps;
        if(curr<0){
            curr = 0;
            return h[0];
        }
        else
        return h[curr];
    }
    
    string forward(int steps) {
        curr+=steps;
        if(curr>h.size()-1)
        {
            curr = h.size()-1;
            return h[h.size()-1];
        }
        else
        return h[curr];
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1582-design-browser-history/design-browser-history.js
/**
 * @param {string} homepage
 */
var BrowserHistory = function(homepage) {
    this.history = [homepage];
    this.curr =  0;
    
};

/** 
 * @param {string} url
 * @return {void}
 */
BrowserHistory.prototype.visit = function(url) {
    this.history = this.history.slice(0,this.curr+1);
    this.history.push(url);
    this.curr++;
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.back = function(steps) {
    this.curr = Math.max(0,this.curr-steps);
    return this.history[this.curr];
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.forward = function(steps) {
    this.curr = Math.min(this.history.length-1,this.curr+steps);
    return this.history[this.curr];
    
};

/** 
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */